bundles/org.openhab.binding.caddx/src/main/java/org/openhab/binding/caddx/internal/CaddxMessage.java

```diff
@@ -0,0 +1,404 @@
+        return checksum1In;
+    }
+
+    public byte getChecksum2In() {
+        return checksum2In;
+    }
+
+    public byte getChecksum1Calc() {
+        return checksum1Calc;
+    }
+
+    public byte getChecksum2Calc() {
+        return checksum2Calc;
+    }
+
+    public CaddxMessageType getCaddxMessageType() {
+        return caddxMessageType;
+    }
+
+    public byte getMessageType() {
+        return message[0];
+    }
+
+    public String getName() {
+        StringBuilder sb = new StringBuilder();
+        sb.append(caddxMessageType.name);
+        switch (caddxMessageType) {
+            case Zone_Status_Request:
+            case Zone_Status_Message:
+                sb.append(" [Zone: ");
+                sb.append(getPropertyById("zone_number"));
+                sb.append("]");
+                break;
+            case Log_Event_Request:
+            case Log_Event_Message:
+                sb.append(" [Event: ");
+                sb.append(getPropertyById("panel_log_event_number"));
+                sb.append("]");
+                break;
+            case Partition_Status_Request:
+            case Partition_Status_Message:
+                sb.append(" [Partition: ");
+                sb.append(getPropertyById("partition_number"));
+                sb.append("]");
+                break;
+            default:
+                break;
+        }
+        return sb.toString();
+    }
+
+    public String getPropertyValue(String property) {
+        if (!propertyMap.containsKey(property)) {
+            logger.debug("Message does not contain property [{}]", property);
+            return "";
+        }
+        return propertyMap.get(property);
+    }
+
+    public String getPropertyById(String id) {
+        if (!idMap.containsKey(id)) {
+            logger.debug("Message does not contain id [{}]", id);
+            return "";
+        }
+        return idMap.get(id);
+    }
+
+    public int @Nullable [] getReplyMessageNumbers() {
+        return caddxMessageType.replyMessageNumbers;
+    }
+
+    public CaddxSource getSource() {
+        return getCaddxMessageType().source;
+    }
+
+    public boolean isChecksumCorrect() {
+        return checksum1In == checksum1Calc && checksum2In == checksum2Calc;
+    }
+
+    public boolean isLengthCorrect() {
+        return message.length == caddxMessageType.length;
+    }
+
+    public boolean hasAcknowledgementFlag() {
+        return hasAcknowledgementFlag;
+    }
+
+    public byte[] getMessageFrameBytes(CaddxProtocol protocol) {
+        if (protocol == CaddxProtocol.Binary) {
+            return getMessageFrameBytesInBinary();
+        } else {
+            return getMessageFrameBytesInAscii();
+        }
+    }
+
+    public byte[] getMessageBytes() {
+        return message;
+    }
+
+    /**
+     * Returns a string representation of a CaddxMessage.
+     *
+     * @return CaddxMessage string
+     */
+    @Override
+    public String toString() {
```

Comment:
There is a lot of code that went into this method. Can you show me where this method would be called?

Reply:
The toString() method is used when logging to get a human interpretation of the byte array.
It is implicitly called from the handlers. For example ThingHandlerZone.java line:147 `logger.trace("caddxEventReceived(): Event Received - {}.", event);`

log excerpt:
<pre>2020-03-24 12:47:44.425 [TRACE] [dx.internal.handler.ThingHandlerZone] - caddxEventReceived(): Event Received - org.openhab.binding.caddx.internal.CaddxEvent[source=Message:  4 Zone Status Message
        Message number:  4 - 4 - -
        Zone number:  2 - 2 - -
        Partition 1 enable: true
        Partition 2 enable: false
        Partition 3 enable: false
        Partition 4 enable: false
        Partition 5 enable: false
        Partition 6 enable: false
        Partition 7 enable: false
        Partition 8 enable: false
        Fire: false
        24 Hour: false
        Key-switch: false
        Follower: true
        Entry / exit delay 1: true
        Entry / exit delay 2: false
        Interior: true
        Local only: false
        Keypad sounder: true
        Yelping siren: true
...
</pre>

Reply:
Resolving this conversation